### PR TITLE
Alert for content editors on incorrect block usage #539

### PR DESCRIPTION
**Pull request by cogniSyb**
Fix #539

**Test Pages:**
Legacy page with V2 Metadata: [BEFORE](https://main--vg-volvotrucks-us--hlxsites.aem.page/drafts/shomps/legacy-with-v2) **|**  [AFTER](https://539-validate-code-sidekick--vg-volvotrucks-us--hlxsites.aem.page/drafts/shomps/legacy-with-v2)

Redesign page with V2 blocks: [BEFORE](https://main--vg-volvotrucks-us--hlxsites.aem.page/drafts/shomps/redesign-with-v1) **|**  [AFTER](https://539-validate-code-sidekick--vg-volvotrucks-us--hlxsites.aem.page/drafts/shomps/redesign-with-v1)

Test URLs:
- Before: https://main--vg-volvotrucks-us--hlxsites.aem.page/
- After: https://539-validate-code-sidekick--vg-volvotrucks-us--hlxsites.aem.page/

This branch contains a debugging script that will trigger on a domain containing hlx, aem, or localhost only when the Sidekick extension is activated. Code will be loaded conditionally as much as possible, and thus won’t trigger for regular visitors.

The configuration is available in constants.xlsx, so that the allow lists for classes and blocks can be updated if necessary. We should be aware that this list needs to be in sync with new code, just like placeholder.xlsx.

The notification is based on the [Snackbar component of Volvo’s design system](https://www.volvogroup.com/experiencesystem/en/#/brand/615bfcc428c818001105b4e3/61a0de32a7047974164f01dd/6230a5b437a675408d86d8cc). It’s a common block, has been created first for Mack Trucks.

To do: add list for (not) allowed items

### CODE REVIEW

**Review by jonatan-lledo-netcentric — DISMISSED** [review]

**Review by alexiscoelho — DISMISSED** [review]

File: scripts/validate-elements.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+const formattedV2AllowedBlocks = formatStringToArray(v2AllowedBlocks);
+
+const allowlist = [''];
+
+const isV2 = document.documentElement.classList.contains('redesign-v2');
+const pageText = isV2 ? 'redesign page' : 'legacy page';
+const targetNode = document.body;
+const config = { childList: true, subtree: false };
+
+const countOccurrences = (elements, getName) => {
```
Comment (jonatan-lledo-netcentric) [inline scripts/validate-elements.js:30]: this code could be more efficient and easy to read:
```
const countOccurrences = (elements, getName) => {
  return elements.reduce((counts, element) => {
    const name = getName(element);
    if (name) {
      counts[name] = (counts[name] ?? 0) + 1;
    }
    return counts;
  }, {});
};
```

File: scripts/validate-elements.js [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  const checkFn = (blockName) => {
+    if (isV2) {
+      return !blockName.startsWith('v2-') && !formattedV2AllowedBlocks.includes(blockName);
+    }
+    return blockName.startsWith('v2-') && !formattedV1AllowedBlocks.includes(blockName);
+  };
+  return collectViolations(blockCounts, checkFn);
+};
+
+const checkSections = () => {
```
Comment (jonatan-lledo-netcentric) [inline scripts/validate-elements.js:67]: same as above
```
const checkSections = () => {
  const sections = targetNode.querySelectorAll(':scope > main > .section');
  const checkList = isV2 ? formattedV1SectionClasses : formattedV2SectionClasses;

  const sectionCounts = Array.from(sections).reduce((counts, section) => {
    const sectionClasses = Array.from(section.classList);

    sectionClasses.forEach((className) => {
      if (checkList.includes(className)) {
        const cleanedClassName = className.startsWith('section--') ? className.replace('section--', '') : className;
        counts[cleanedClassName] = (counts[cleanedClassName] ?? 0) + 1;
      }
    });

    return counts;
  }, {});

  const checkFn = (className) => !allowlist.includes(className);
  return collectViolations(sectionCounts, checkFn);
};
```

File: scripts/validate-elements.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+  const blockViolations = checkBlocks();
+  const sectionViolations = checkSections();
+  logCombinedViolations(blockViolations, sectionViolations);
+};
+
+const existingElement = document.querySelector('body > helix-sidekick');
+if (existingElement) {
+  performDomCheck();
+} else {
+  const callback = (mutationsList, observer) => {
```
Comment (jonatan-lledo-netcentric) [inline scripts/validate-elements.js:119]: try this simplified version of the same piece of code
```
const callback = (mutationsList, observer) => {
  for (const mutation of mutationsList) {
    if (mutation.type !== 'childList') continue;

    for (const node of mutation.addedNodes) {
      if (node.nodeType !== 1 || node.tagName.toLowerCase() !== 'helix-sidekick') continue;

      performDomCheck();
      observer.disconnect();
      return;
    }
  }
};
```

File: constants.json [hunk 4]
```diff
@@ -76,21 +76,6 @@
       }
     ]
   },
-  "headerConfig": {
```
[the remark this answers is not in the document]
Comment (alexiscoelho) [reply constants.json:79]: It was duplicated

[the remark this answers is not in the document]
Comment (santi-homps) [reply constants.json:79]: Precisely

File: scripts/validate-elements.js (hunk 2 above)
Comment (santi-homps) [reply scripts/validate-elements.js:67]: This brings unexpected linter issues

File: scripts/validate-elements.js (hunk 3 above)
Comment (santi-homps) [reply scripts/validate-elements.js:119]: This brings unexpected linter issues

File: constants.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (jonatan-lledo-netcentric) [reply constants.json:79]: so it's ok then

File: scripts/validate-elements.js (hunk 1 above)
Comment (jonatan-lledo-netcentric) [reply scripts/validate-elements.js:30]: no worries then. don't touch it

File: scripts/validate-elements.js (hunk 2 above)
Comment (jonatan-lledo-netcentric) [reply scripts/validate-elements.js:67]: I just try to reduce the amount of inner scopes

File: scripts/validate-elements.js (hunk 3 above)
Comment (jonatan-lledo-netcentric) [reply scripts/validate-elements.js:119]: let's left it as it is for consistency then

**Review by jonatan-lledo-netcentric — APPROVED** [review]

**Review by alexiscoelho — APPROVED** [review]

**Review by amglebihan — APPROVED** [review]

**Pull request merged**